pkg/reconcilermanager/controllers/reconciler_finalizer_handler.go

```diff
@@ -0,0 +1,94 @@
+)
+
+// reconcilerFinalizerHandler is a delegator that handles the
+// `configsync.gke.io/reconciler` finalizer on behalf of the reconciler.
+//
+// The reconciler adds the finalizer when deletion propagation is set to
+// `Foreground`, and removes it when R*Sync has a deletionTimestamp.
+// If the reconciler is unable to update R*Sync to remove the finalizer, R*Sync
+// would get stuck in the deletion process.
+// To unblock the deletion, the reconciler-manager acts as a delegator to remove
+// R*Sync's finalizer when needed.
+type reconcilerFinalizerHandler interface {
+	// handleReconcilerFinalizer handles the `configsync.gke.io/reconciler`
+	// finalizer when the reconciler is not able to remove it from R*Sync.
+	handleReconcilerFinalizer(context.Context, client.Object) (bool, error)
+}
+type rootReconcilerFinalizerHandler struct {
+}
+
+// handleReconcilerFinalizer waits for the root reconciler finalizer to finish.
+// Note: it is possible RootSync is stuck in deletion because of the finalizer
+// (e.g. root-reconciler is either deleted, or unable to remove the finalizer),
+// but we decided to leave it as a known issue and brainstorm a better solution later.
+func (r rootReconcilerFinalizerHandler) handleReconcilerFinalizer(context.Context, client.Object) (bool, error) {
+	// Waiting for the root reconciler finalizer to finish
+	return false, nil
+}
+
+type nsReconcilerFinalizerHandler struct {
+	loggingController
+	client client.Client
+}
+
+// handleReconcilerFinalizer removes the 'configsync.gke.io/reconciler'
+// finalizer from the RepoSync object when the RepoSync Namespace is terminating.
+//
+// When the RepoSync's Namespace is terminating, Namespace controller will
+// handle the deletion propagation of all resources managed by the RepoSync.
+// Therefore, remove the RepoSync finalizer to avoid duplicated work.
+// This also fixes an issue with a race when Namespace controller deletes
+// the reconciler's RoleBinding needed for removing the finalizer.
+//
+// It returns whether the RepoSync finalizer has been removed.
+func (r nsReconcilerFinalizerHandler) handleReconcilerFinalizer(ctx context.Context, syncObj client.Object) (bool, error) {
+	syncObjNamespace := &corev1.Namespace{}
+	syncObjNamespace.Name = syncObj.GetNamespace()
+	key := client.ObjectKeyFromObject(syncObjNamespace)
+	if err := r.client.Get(ctx, key, syncObjNamespace); err != nil {
+		r.logger(ctx).Error(err, "Get sync object Namespace failed")
+		return false, err
+	}
+
+	if syncObjNamespace.Status.Phase != corev1.NamespaceTerminating {
+		// RepoSync's namespace is active, wait for the reconciler finalizer to complete.
+		return false, nil
+	}
+
+	// RepoSync's namespace is terminating, remove the reconciler finalizer.
+	if !controllerutil.RemoveFinalizer(syncObj, metadata.ReconcilerFinalizer) {
+		r.logger(ctx).Info("The reconciler finalizer is not found")
+		return true, nil
+	}
+	if err := r.client.Update(ctx, syncObj); err != nil {
```

Comment:
Would it make more sense to patch the finalizer removal?

Erroring on update conflict and retrying might be ok, but patching seems potentially faster.

Reply:
I see it was already updating. So that's fine. Nevermind.

Reply:
I used a consistent style as removing the reconciler-manager finalizer: https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/reconcilermanager/controllers/reconciler_base.go#L640.

Agreed that patching might be more efficient. We can change them altogether in a separate PR without any functionality change.